Proxy model is not updated issue fixed by adding _get_fields_of_sende…

firtk opened the pull request:
When fields are registered proxy models are not included inside the registry dictionary. So this line always fails `if sender in registry:` for a proxy models. I added the `_get_fields_of_sender` function. If it can't find the model inside registry dictionary, It will second check the base classes of the model, If base class found, it will get the field from that and returns the fields.

firtk committed:
Proxy model is not updated issue fixed by adding _get_fields_of_sender function

st4lk commented on audit_log/middleware.py:
```diff
@@ -33,6 +33,19 @@ def _get_user_from_request(request):
+            # Check it's base classes.
+            for base in sender.__bases__:
+                if base in registry:
+                    fields = registry.get_fields(base)
```
Shouldn't we break the for-loop here?
Or we want to get the last possible base?

firtk replied:
I checked in a test project. Yes we will break there. I will create another commit

firtk committed:
return statement added in a for loop

st4lk reviewed (commented):
Thanks @firtk !
Can you please add at least one test to check this case?

firtk commented:
Sure. If i can, i will add it tomorrow.

firtk committed:
Tests are updated for django 2.2.24 and Proxy Model Test added

firtk commented:
Hi @st4lk,

The tests were for Django 1.7, I guess. I did use Django 2.2.24 as we use and updated some configurations for it. Changes I made :
- Added on_delete to necessary fields for tests/audit_log_tests/models.py
- Moved urlpatterns from test files to test/urls.py, because I haven't been able to set it up like that.
- Changed urls to re_path's.
- Changed some settings inside runtests.py for the Django 2.2
- Changed some relative imports to absolute imports, because i was getting errors.

All previous tests are running successfully and I added the TestProxyModel case.

How i runned the tests:
- Cloned Repo
- Installed Django 2.2.24
- Run `python setup.py develop`
- Run `runtests.py`

I couldn't manage to run `runtests_custom_auth.py` . I guess, with this Test setup Django couldn't figure out some settings. If I figure out I will inform you or can you point me to right direction ?

chapayevdauren commented on audit_log/tests/audit_log_tests/models.py:
```diff
@@ -5,7 +5,10 @@
 )
 
-from audit_log.models.fields import LastUserField, LastSessionKeyField, CreatingUserField
+from audit_log.models.fields import (LastUserField,
+                                    LastSessionKeyField,
+                                    CreatingUserField,
+                                    CreatingSessionKeyField)
```
```suggestion
from audit_log.models.fields import (
    LastUserField,
    LastSessionKeyField,
    CreatingUserField,
    CreatingSessionKeyField,
)
```

chapayevdauren replied:
sorry, ignore it for now

chapayevdauren commented on audit_log/tests/runtests.py:
```diff
@@ -78,14 +85,14 @@
+args.append("--verbosity=2")
 args.extend(sys.argv[offset:])
-
+print(args)
```
👮‍♂️

firtk replied:
I missed that 😄

firtk committed:
Removed forgotten print statement

st4lk commented on audit_log/tests/audit_log_tests/models.py:
```diff
@@ -5,7 +5,10 @@
 )
 
-from audit_log.models.fields import LastUserField, LastSessionKeyField, CreatingUserField
+from audit_log.models.fields import (LastUserField,
+                                    LastSessionKeyField,
+                                    CreatingUserField,
+                                    CreatingSessionKeyField)
```
Propose to not introduce any code-style refactoring in this PR

st4lk commented on audit_log/tests/audit_log_tests/test_logging.py:
```diff
@@ -235,3 +215,44 @@ def test_logging(self):
+        _setup_admin()
+        # Create Normal Item with user 1
+        client_1 = Client()
+        client_1.login(username = "admin@example.com", password = "admin")
```
Can we please be consistent - always use single quotes (`'`), not double quotes?

st4lk commented on audit_log/tests/runtests.py:
```diff
@@ -8,68 +8,75 @@
+    "django.contrib.staticfiles",
+]
 
-ALWAYS_MIDDLEWARE_CLASSES = (
-    'django.contrib.sessions.middleware.SessionMiddleware',
-    'django.middleware.common.CommonMiddleware',
-    'django.middleware.csrf.CsrfViewMiddleware',
-    'django.contrib.auth.middleware.AuthenticationMiddleware',
-    'django.contrib.messages.middleware.MessageMiddleware',
-    'django.middleware.clickjacking.XFrameOptionsMiddleware',
-    'audit_log.middleware.UserLoggingMiddleware',
-)
+ALWAYS_MIDDLEWARE_CLASSES = [
+    "django.contrib.sessions.middleware.SessionMiddleware",
+    "django.middleware.common.CommonMiddleware",
+    "django.middleware.csrf.CsrfViewMiddleware",
+    "django.contrib.auth.middleware.AuthenticationMiddleware",
+    "django.contrib.messages.middleware.MessageMiddleware",
+    "django.middleware.clickjacking.XFrameOptionsMiddleware",
+    "audit_log.middleware.UserLoggingMiddleware",
+]
```
Propose to not introduce any code-style refactoring in this PR (i.e. keep single quotes)

st4lk commented on audit_log/tests/runtests.py:
```diff
@@ -8,68 +8,75 @@
 django.setup()
-args = [sys.argv[0], 'test']
+args = [sys.argv[0], "test"]
 # Current module (``tests``) and its submodules.
-test_cases = '.'
+test_cases = "."
```
Propose to not introduce any code-style refactoring in this PR - use single quotes

st4lk commented on audit_log/tests/runtests.py:
```diff
@@ -78,14 +85,13 @@
     pass
 else:
-    option = sys.argv[1].startswith('-')
+    option = sys.argv[1].startswith("-")
```
Propose to not introduce any code-style refactoring in this PR

st4lk commented on audit_log/tests/runtests.py:
```diff
@@ -78,14 +85,13 @@
 args.append(test_cases)
 # ``verbosity`` can be overwritten from command line.
-args.append('--verbosity=2')
+args.append("--verbosity=2")
```
Propose to not introduce any code-style refactoring in this PR

st4lk reviewed (approved):
LGTM with comments.
Let's focus on one thing in PR. And not introduce unrelated refactoring (like changing import styles or quote styles).
We can do it later if needed - but we need certain rules for that.

firtk committed:
Code-style refactorings removed

firtk commented:
I reverted automatic code refactoring. Can you create a new version if everything is okay?

By the way, I managed to run `runtests_custom_auth.py` with the below settings. I am not including those in this PR because when we use them `runtests.py` will fail.

- Make `runtests_custom_auth.py` same as 'runtests.py'

    - Change => `CUSTOM_INSTALLED_APP = ('audit_log_tests')`
    - Add to settings => `AUTH_USER_MODEL = "audit_log_tests.Employee",`

- In audit_log_tests/views.py.

    - Change => `from audit_log.tests.audit_log_tests.models ...` to `from audit_log_tests.models ...`

- In audit_log_tests/test_logging.py.

    - Change all => `from .models ...` to `from audit_log_tests.models ...`

- In audit_log_tests/test_manager.py.

    - Change => `from .models ...` to `from audit_log_tests.models ...`

With these changes `runtests_custom_auth.py` will run. But TrackingChangesTest.test_logging_custom_user is failing. This issue is related to [Issue 24](https://github.com/vvangelovski/django-audit-log/issues/24) and still not resolved I guess.
